### PR TITLE
Reimplement actor isolation checking for referencing a declaration.

### LINKED ISSUE #58107: [SR-15833] Compiler error when trying to implement protocol with async methods in actor
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15833      |
   |Radar             | None         |
   |Original Reporter | itwenty (JIRA User)      |
   |Type              | Bug    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler, swift    |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: b61d4557d94dccf0825d904bfee2f154

  </details>

**Issue Description:**

Given this protocol

``` swift
protocol ApodFetcher: AnyObject {
    func fetchApod(date: Date) async throws -> Apod
}
```

and this actor

``` swift
actor ApodApiFetcher: ApodFetcher {
    func fetchApod(date: Date) throws -> Apod {
        ...
    }
}
```

the compiler complains "Actor-isolated instance method 'fetchApod(date: )' cannot be used to satisfy a protocol requirement". [The actors proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0306-actors.md#protocol-conformance) mentions this to be a valid scenario, so not sure why I get this error.

Running Xcode 13.2.1 on macOS 12.1. The project was iOS with deployment target iOS 15.1

### PR DESCRIPTION
**Pull request by DougGregor**
Start collapsing the several implementations of actor isolation checking
into a single place that determines what it means to reference a declaration
from a given context, potentially supplying an instance for an actor. This
is partly cleanup, and partly staging for the implementation of the
Sendable restrictions introduced in SE-0338. The result of this check
falls into one of three categories:

* Reference occurs within the same concurrency domain (actor/task)
* Reference leaves an actor context to a nonisolated context (SE-0338)
* Reference enters the context of the actor, which might require a
combination of implicit async, implicit throws, and a "distributed" check.

Throughout this change I've sought to maintain the existing semantics,
even where I believe they are incorrect. The changes to the test cases
are not semantic changes, but reflect the unification of some
diagnostic paths that changed the diagnostic text but not when or how
those diagnostics are produced. Additionally, SE-0338 has not yet been
implemented, although this refactoring makes it easier to implement
SE-0338.

Use this new actor isolation checking scheme to implement the most
common actor-isolation checks for accessing both member and
non-member declarations from expressions.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -9352,3 +9352,7 @@ void swift::simple_display(llvm::raw_ostream &out, AnyFunctionRef fn) {
   else
     out << "closure";
 }
+
+bool ActorIsolation::isDistributedActor() const {
+  return getKind() == ActorInstance && getActor()->isDistributedActor();
+}
```
Comment (ktoso) [inline lib/AST/Decl.cpp:9358]: ❤️

File: lib/Sema/TypeCheckConcurrency.h [hunk 2]
```diff
@@ -222,6 +222,151 @@ bool contextRequiresStrictConcurrencyChecking(
...
+  /// of a member access or a parameter passed to a function.
+  static ActorReferenceResult forReference(
+      ConcreteDeclRef declRef,
+      SourceLoc declRefLoc,
+      const DeclContext *fromDC,
+      Optional<VarRefUseEnv> useKind,
+      Optional<ReferencedActor> actorInstance = None);
+
+  operator Kind() const { return kind; }
+};
```
Comment (ktoso) [inline lib/Sema/TypeCheckConcurrency.h:368]: This entire way to model it is excellent, really looking forward to it 👍

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/Sema/CSApply.cpp [hunk 3]
```diff
@@ -7821,55 +7809,60 @@ bool ExprRewriter::isDistributedThunk(ConcreteDeclRef ref, Expr *context) {
...
+      ref, context->getLoc(), referenceDC, None, actorRef);
+  switch (refResult) {
+  case ActorReferenceResult::ExitsActorToNonisolated:
+  case ActorReferenceResult::SameConcurrencyDomain:
     return false;
+
+  case ActorReferenceResult::EntersActor:
+    return refResult.options.contains(ActorReferenceResult::Flags::Distributed);
   }
 }
```
Comment (ktoso) [inline lib/Sema/CSApply.cpp:7867]: This is awesome, thanks for centralizing all the logic 🥳

**Pull request merged**